### PR TITLE
add Z between types

### PR DESCRIPTION
**Pull request by 3pmile**
**Description**

This PR implements...
- [x] the Add trait for Z + Zq
- [x] the Add trait for Z + Q
- [x] the Add trait for Zq and Standard types u64, u32, u16, u8, i64, i32, i16, i8

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/integer/z/arithmetic/add.rs [hunk 1]
```diff
@@ -144,6 +230,70 @@ mod test_add_between_types {
...
+            Q::from_str(&format!("1/{}", u64::MAX)).unwrap()
+                + Q::from_str(&format!("{}/1", u64::MAX)).unwrap()
+        );
+        assert_eq!(
+            e,
+            Q::from_str(&format!("{}/1", u64::MAX)).unwrap()
+                + Q::from_str(&format!("{}/2", u64::MAX)).unwrap()
+        );
+    }
+}
```
Comment (Marcel583) [inline src/integer/z/arithmetic/add.rs:295]:
```suggestion
    /// testing addition for big numbers
    #[test]
    fn add_large_numbers() {
        let a: Z = Z::from(u64::MAX);
        let b: Q = Q::from_str(&format!("1/{}", u64::MAX)).unwrap();
        let c: Q = Q::from_str(&format!("{}/2", u64::MAX)).unwrap();
        
        let d: Q = &a + b;
        let e: Q = a + c;
        
        assert_eq!(
            d,
            Q::from_str(&format!("1/{}", u64::MAX)).unwrap()
                + Q::from_str(&format!("{}/1", u64::MAX)).unwrap()
        );
        assert_eq!(
            e,
            Q::from_str(&format!("{}/1", u64::MAX)).unwrap()
                + Q::from_str(&format!("{}/2", u64::MAX)).unwrap()
        );
    }
}
```

File: src/integer/z/arithmetic/add.rs [hunk 2]
```diff
@@ -197,3 +347,62 @@ mod test_add {
...
+        let c: Zq = Zq::try_from((i64::MAX - 1, i64::MAX)).unwrap();
+
+        let d: Zq = &a + b;
+        let e: Zq = a + c;
+        assert_eq!(
+            d,
+            Zq::try_from(((u64::MAX - 1) / 2 + 58, u64::MAX - 58)).unwrap()
+        );
+        assert_eq!(e, Zq::try_from((0, i64::MAX)).unwrap());
+    }
```
Comment (Marcel583) [inline src/integer/z/arithmetic/add.rs:407]:
```suggestion
    /// testing addition for big numbers
    #[test]
    fn add_large_numbers() {
        let a: Z = Z::from(u64::MAX);
        let b: Zq = Zq::try_from((i64::MAX, u64::MAX - 58)).unwrap();
        let c: Zq = Zq::try_from((i64::MAX - 1, i64::MAX)).unwrap();

        let d: Zq = &a + b;
        let e: Zq = a + c;
        
        assert_eq!(
            d,
            Zq::try_from(((u64::MAX - 1) / 2 + 58, u64::MAX - 58)).unwrap()
        );
        assert_eq!(e, Zq::try_from((0, i64::MAX)).unwrap());
    }
```

**Review by Marcel583 — APPROVED** [review]

File: src/integer/z/arithmetic/add.rs [hunk 3]
```diff
@@ -51,6 +59,84 @@ arithmetic_trait_borrowed_to_owned!(Add, add, Z, Z, Z);
...
+        let mut out = fmpz(0);
+        unsafe {
+            fmpz_mod_add_fmpz(
+                &mut out,
+                &other.value.value,
+                &self.value,
+                &*other.modulus.modulus,
+            );
+        }
+        Zq::from_z_modulus(&Z::from_fmpz(&out), &other.modulus)
```
Comment (Marcel583) [inline src/integer/z/arithmetic/add.rs:133]: see: mul Zq between types
